TLM/TLM/Patch/NetManagerEvents.cs

```diff
@@ -0,0 +1,98 @@
+    /// The present participle (e.g. <see cref="ReleasingLane"/>) identifies a presently/imminently
+    /// occurring event--that is, one that is about to or is in the process of happening. The past
+    /// simple (e.g. <see cref="ReleasedLane"/>) identifies an event that is already completed. In
+    /// most cases, these respectively correspond to prefix and postfix Harmony patches.
+    /// </para>
+    /// </remarks>
+    [HarmonyPatch(typeof(NetManager))]
+    internal class NetManagerEvents {
+
+        internal delegate void ReleasingLaneEventHandler(uint laneId, ref NetLane lane);
+        internal delegate void ReleasedLaneEventHandler(uint laneId);
+
+        internal delegate void ReleasingSegmentEventHandler(ushort segmentId, ref NetSegment segment);
+        internal delegate void ReleasedSegmentEventHandler(ushort segmentId);
+
+        private NetManagerEvents() {
+        }
+
+        /// <summary>
+        /// Occurs when a lane is about to be released.
+        /// In most cases the lane is still valid when this event occurs, so that <see cref="NetLaneExtensions.IsValidWithSegment(ref NetLane)"/> will return true.
+        /// </summary>
+        public event ReleasingLaneEventHandler ReleasingLane;
+
+        /// <summary>
+        /// Occurs when a lane has been released and is no longer valid.
+        /// </summary>
+        public event ReleasedLaneEventHandler ReleasedLane;
+
+        /// <summary>
+        /// Occurs when a segment is about to be released.
+        /// In most cases the segment is still valid when this event occurs, so that <see cref="NetSegmentExtensions.IsValid(ref NetSegment)"/> will return true.
+        /// </summary>
+        public event ReleasingSegmentEventHandler ReleasingSegment;
+
+        /// <summary>
+        /// Occurs when a segment has been released and is no longer valid.
+        /// </summary>
+        public event ReleasedSegmentEventHandler ReleasedSegment;
+
+        public static NetManagerEvents Instance { get; } = new NetManagerEvents();
+
+        [HarmonyPostfix]
+        [HarmonyPatch("ReleaseLaneImplementation")]
+        [UsedImplicitly]
+        internal static void ReleaseLaneImplementationPostfix(uint lane) => Instance.ReleasedLane?.Invoke(lane);
+
+        [HarmonyPostfix]
+        [HarmonyPatch(
+            "ReleaseSegmentImplementation",
+            new[] { typeof(ushort), typeof(NetSegment), typeof(bool) },
+            new[] { ArgumentType.Normal, ArgumentType.Ref, ArgumentType.Normal }
+        )]
+        [UsedImplicitly]
+        internal static void ReleaseSegmentImplementationPostfix(ushort segment) => Instance.ReleasedSegment?.Invoke(segment);
+
+        [HarmonyPrefix]
+        [HarmonyPatch(
+            "PreReleaseSegmentImplementation",
+            new[] { typeof(ushort), typeof(NetSegment), typeof(bool) },
+            new[] { ArgumentType.Normal, ArgumentType.Ref, ArgumentType.Normal }
+        )]
```

Comment:
[Note]
maybe you can create this attribute:
```cs
    public class HarmonyPatch2 : HarmonyLib.HarmonyPatch {
        public HarmonyPatch2(Type delcaringType, Type delegateType, bool instance = false){
            info.declaringType = delcaringType;
            info.methodName = delegateType.Name;
            var args = delegateType.GetMethod("Invoke").GetParameters().Select(p => p.ParameterType); // or use what we have in transpiler utils.
            if (instance) args = args.Skip(1); // skip arg0 because its instance method.
            info.argumentTypes = args.ToArray();
        }
    }
```
And then use it like this:
```cs
private delegate void PreReleaseSegmentImplementation(ushort segment, ref NetSegment data, bool keepNodes);

[HarmonyPatch2(typeof(NetManager), typeof(PreReleaseSegmentImplementation))]
````
The advantage of doing it this way is that you are less likely to make mistake copy pasting the function from dnSpy.

Reply:
I wasn't even using dnSpy. I did this all from within Visual Studio. But I'm still enough of a novice with Harmony that I'd rather wait to do things like this until I have a better handle on the big picture of how it works.